lib/BPFBase.ml

```diff
@@ -0,0 +1,422 @@
+(*                           the diy toolsuite                              *)
+(*                                                                          *)
+(* Copyright (c) 2024 Puranjay Mohan <puranjay@kernel.org>                  *)
+(*                                                                          *)
+(*                                                                          *)
+(* This software is governed by the CeCILL-B license under French law and   *)
+(* abiding by the rules of distribution of free software. You can use,      *)
+(* modify and/ or redistribute the software under the terms of the CeCILL-B *)
+(* license as circulated by CEA, CNRS and INRIA at the following URL        *)
+(* "http://www.cecill.info". We also give a copy in LICENSE.txt.            *)
+(****************************************************************************)
+
+(** BPF architecture, base definitions *)
+
+open Printf
+open Sign
+
+let arch = Archs.bpf
+let endian = Endian.Little
+let base_type = CType.Base "int"
+
+(*************)
+(* Registers *)
+(*************)
+
+type ireg =
+  | R0
+  | R1
+  | R2
+  | R3
+  | R4
+  | R5
+  | R6
+  | R7
+  | R8
+  | R9
+  | R10
+
+type reg =
+  | IReg of ireg
+  | PC
+  | Symbolic_reg of string
+  | Internal of int
+
+let parse_ireg = function
+  | "R0" | "r0" -> R0
+  | "R1" | "r1" -> R1
+  | "R2" | "r2" -> R2
+  | "R3" | "r3" -> R3
+  | "R4" | "r4" -> R4
+  | "R5" | "r5" -> R5
+  | "R6" | "r6" -> R6
+  | "R7" | "r7" -> R7
+  | "R8" | "r8" -> R8
+  | "R9" | "r9" -> R9
+  | "R10" | "r10" | "FP" | "fp" -> R10
+  | _ -> raise Exit
+;;
+
+let parse_reg s =
+  try Some (IReg (parse_ireg s)) with
+  | Exit -> None
+;;
+
+open PPMode
+
+let do_pp_ireg = function
+  | R0 -> "r0"
+  | R1 -> "r1"
+  | R2 -> "r2"
+  | R3 -> "r3"
+  | R4 -> "r4"
+  | R5 -> "r5"
+  | R6 -> "r6"
+  | R7 -> "r7"
+  | R8 -> "r8"
+  | R9 -> "r9"
+  | R10 -> "fp"
+;;
+
+let pp_reg = function
+  | IReg r -> do_pp_ireg r
+  | Symbolic_reg r -> r
+  | Internal i -> sprintf "ir%i" i
+  | PC -> "pc"
+;;
+
+let reg_compare = compare
+
+let symb_reg_name = function
+  | Symbolic_reg s -> Some s
+  | _ -> None
+;;
+
+let symb_reg r = Symbolic_reg r
+let type_reg _ = base_type
+
+(************)
+(* Barriers *)
+(************)
+
+(* Artificial sync instruction for testing *)
+type barrier = Sync
+
+let all_kinds_of_barriers = [ Sync ]
+
+let pp_barrier = function
+  | Sync -> "Sync"
+;;
+
+let barrier_compare = compare
+
+(***************)
+(* Annotations *)
+(***************)
+
+(* X  -> Atomic without any ordering *)
+(* SC -> Atomic with full ordering *)
+(* N  -> Plain: non-atomic and no ordering *)
+type lannot =
+  | X
+  | SC
+  | N
+
+(****************)
+(* Instructions *)
+(****************)
+
+type k = int
+type lbl = Label.t
+
+type op =
+  | ADD
+  | SUB
+  | MUL
+  | DIV
+  | REM
+  | AND
+  | OR
+  | XOR
+  | LSL
+  | LSR
+  | ASR
+  | AMOXCHG
+  | AMOCMPXCHG
```

Comment:
Wouldn't it be better to have two types for operators, one standard and the the other for AMO's. Even if some operators are common to the two sets (_e.g._ `ADD`) they can be represented by different constructors. One can even have the same name for different constructors by uslng local modules:
```OCaml
module Std = struct
  type op = ... | ADD | ...
end

module AMO = struct
  type op.= ... | ADD | ...
end

(* Here refer to `AMO.ADD` or `Std.ADD` *)

type instruction =
| Op of Std.op * ...
| ...
| AMO of AMO.op * ...
| ... 
```
That way, typing controls more and we avoid some `assert false`.

Reply:
Sure,
I will fix this in the next version

Reply:
I tried this but it is making the implementation difficult.
In the Lexer, `+=` can mean both `Std.ADD` and `AMO.ADD`. To make this work I will have to change a lot of code. So, if this is not a blocker, I would like to keep the current implementation.

Reply:
Ok let it as it is.  Out of curiosity, how is disambiguation peformed?

Reply:
Ok, let us leave it that way for now.

Reply:
The disambiguation is performed in the Parser as the normal atomic ops look like `r1 += r2` and the atomic ops look like `lock *(u64 *)(r1 + 32) += r2`